File: AI_Engine_Development/AIE/Feature_Tutorials/26-gentle_intro_to_aie_kernel_programming/unit_tests/sqmag/src/sqmag.cpp
/********************************************************************
* Copyright © 2025 Advanced Micro Devices, Inc. All rights reserved.
* SPDX-License-Identifier: MIT
********************************************************************/

#include "aie_api/aie.hpp"
#include "aie_api/aie_adf.hpp"

#include "sqmag.hpp"

//#define KDEBUG	// comment out to suppress debug messages

#ifdef KDEBUG
	#include "aie_api/utils.hpp"
#endif // KDEBUG

using namespace aie;

// calculate the squared magnitude of the input vector elements
// processes "burst_count" vectors during each invocation
template  <typename Ti, typename To, unsigned vlen, unsigned burst_count>
void sqmag(input_stream<Ti>  *istrm,	// input stream
           output_stream<To> *ostrm		// output stream
) {

	for (auto i = 0u; i < burst_count; i++) {

        auto vin = readincr_v<vlen>(istrm);	// read "vlen" samples from "istrm" (re + j*im)
                                        	// into the vector register "vin"
        auto magsq = abs_square<To>(vin);  	// use "abs_square()" API to calculate (re^2 + im^2)
                                            // for each element and store in vector register "magsq"

		#ifdef KDEBUG
			printf("i = %d:\n", i);
			print(vin, true, "  vin = ");
			print(magsq, true, "  magsq = ");
			printf("\n");
		#endif // KDEBUG

		writeincr(ostrm, magsq);    // write vector register "magsq"
                                    // to output stream "ostrm"

	} // end for (auto i = 0u; i < burst_cnt; i++)

} // end sqmag()

File: AI_Engine_Development/AIE/Feature_Tutorials/26-gentle_intro_to_aie_kernel_programming/unit_tests/sqmag/julia/sqmag.jl
#====================================================================
  Copyright © 2025 Advanced Micro Devices, Inc. All rights reserved.
  SPDX-License-Identifier: MIT
====================================================================#

debug = true    # comment out to disable debug messages

include("../../../contrived/julia/utilities.jl")

# !!! User Parameters !!! #
# !!! Make sure that they match "src/params.h" !!!

vlen = 8     	# vector length
burst_count = 8 # number of "samples" to process per function invocation
iter_count = 2  # no. of function invocations

# !!! End User Parameters !!! #

# !!! Derived Parameters !!! #
N = burst_count * iter_count    # total number of "samples" to generate

# !!! End Derived Parameters !!! #

using Random
Random.seed!(11)    # set seed for repeatability

# create the complex vector
# !!! N.B.: Julia does "no automatic promotion" !!!
x = rand(Complex{Int8}, (vlen, N))	# column-major
x = convert.(Complex{Int16}, x)		# limit the elements to Int8 values

if debug
    @printf("\n")
    @printf("Input vector:\n")
    display(x)
    @printf("\n")
    max_x_re = maximum(real(x))
    max_x_im = maximum(imag(x))
    max_x = (max_x_re > max_x_im) ? max_x_re : max_x_im
    @printf("Maximum x value: %d\n", max_x)
    min_x_re = minimum(real(x))
    min_x_im = minimum(imag(x))
    min_x = (min_x_re < min_x_im) ? min_x_re : min_x_im
    @printf("Minimum x value: %d\n", min_x)
    @printf("\n")
end # if debug

# calculate the squared magnitude
y = abs2.(x)

if debug
    @printf("Output vector:\n")
    display(y)
    @printf("\n")
    max_y = maximum(y)
    min_y = minimum(y)
    @printf("Maximum y value: %d\n", max_y)
    @printf("Minimum y value: %d\n", min_y)
    @printf("\n")
end # if debug

# write results
# !!! Make sure that the PLIO width used matches that in graph.hpp !!!
arr2file(x, false, 64, "ivec.dat")	# AIE expects row-major
arr2file(y, false, 64, "ovec.dat")

run(`mv ivec.dat ../data/.`)	# move "ivec.dat" to the data directory

